### PR TITLE
Show time remaining estimate during conversion

### LINKED ISSUE #52: Show time left estimate during conversion
It might be useful to show an estimate of the time left during the GIF creation. The conversion speed seems pretty stable, at least from my naive perspective.

@kornelski Thoughts on the best way to calculate it?

---

Note, this is not an easy issue. It's expected that you put a lot of work and perfection into your work.

### PR DESCRIPTION
**Pull request by allewun**
Fixes #52

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for working on this 👍

**Comment by sindresorhus** [comment]
Note this from the original issue, which should be handled:

> Yeah, there will be some uneven progress in the beginning (roughly for the first N frames where N = cpu cores), but then it should be smooth.

File: Gifski/MainWindowController.swift [hunk 1]
```diff
@@ -130,6 +141,10 @@ final class MainWindowController: NSWindowController {
 
 	private var progress: Progress?
 
+	private var progressObserver: NSKeyValueObservation?
+	private var indeterminateObserver: NSKeyValueObservation?
+	private var startTime = NSDate()
```
Comment (sindresorhus) [inline Gifski/MainWindowController.swift]: Why not use `Date()`?

**Comment by sindresorhus** [comment]
I don't think we should show `Calculating time remaining…`. Would be better to just wait to fade in the time remaining label when we have a guesstimate.

**Comment by sindresorhus** [comment]
I think we should only show seconds when less than 1 minutes.

File: Gifski/MainWindowController.swift (hunk 1 above)
Comment (allewun) [reply Gifski/MainWindowController.swift]: muscle memory :)

**Comment by allewun** [comment]
> Note this from the original issue, which should be handled:
> 
> > Yeah, there will be some uneven progress in the beginning (roughly for the first N frames where N = cpu cores), but then it should be smooth.

Regarding this, it seems unavoidable that the estimation will be inaccurate at start. Browser file downloads, torrent downloads, video encoding etc.. all tend to behave similarly from what I've seen. We could try waiting a few more seconds before showing the progress so it feels more accurate overall.

**Comment by sindresorhus** [comment]
> We could try waiting a few more seconds before showing the progress so it feels more accurate overall.

Yeah, I think that would be good.

**Comment by sindresorhus** [comment]
And after waiting a few more seconds, if the estimated progress is 10 seconds or less, I don't think we should show the time estimate at all.

**Comment by sindresorhus** [comment]
I think we should fade it out when the time estimate has reached 1 seconds or less, as the below doesn't look very good (It stays there until the progress circle also fades out):

Screen Shot 2019-03-20 at 16 13 12

File: Gifski/MainWindowController.swift [hunk 2]
```diff
@@ -186,6 +217,30 @@ final class MainWindowController: NSWindowController {
...
+		timeRemainingLabel.translatesAutoresizingMaskIntoConstraints = false
+
+		NSLayoutConstraint.activate([
+			timeRemainingLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
+			timeRemainingLabel.topAnchor.constraint(equalTo: circularProgress.bottomAnchor)
+		])
+	}
+
+	private func formattedTimeRemaining(seconds: TimeInterval) -> String? {
+		elapsedTimeFormatter.allowedUnits = seconds < 60 ? .second : .minute
```
Comment (sindresorhus) [inline Gifski/MainWindowController.swift]: It should still show hours.

```suggestion
		elapsedTimeFormatter.allowedUnits = seconds < 60 ? .second : [.hour, .minute]
```

**Comment by allewun** [comment]
@sindresorhus I moved the estimation logic into its own class and implemented:
1. buffer for 3 seconds before showing estimate
2. if initial estimate is less than 10 seconds, don't bother showing it at all
3. fade the estimate 1 second before finishing


**Long processing, show label after estimate stabilizes**:
buffer

**Short processing, don't show estimate**:
skip

**Comment by sindresorhus** [comment]
This is looking great now. Thanks for working on this, @allewun 🙌

**Comment by allewun** [comment]
@sindresorhus thanks for working with me on it!

**Comment by sindresorhus** [comment]
It will be out very soon. It’s waiting for review.

**Comment by geordiemhall** [comment]
Great change! 

Very minor thing - might be nicer to round away from zero for the time shown in the label so that it never hits "0 seconds remaining" (would just remain on "1 second remaining" during fade out).

**Comment by sindresorhus** [comment]
@geordiemhall I agree, and already did that in a follow-up commit: https://github.com/sindresorhus/gifski-app/commit/fc5482ee9c74f26b5281f4c80d848626d8c3dc1e#diff-81f725e4fa6f9c266d8d01bc0c462570R104

**Comment by geordiemhall** [comment]
Ah awesome, didn’t notice sorry!

**Pull request merged**